File: backend/fixed_charges.py
"""Recurring monthly charges applied on top of card statement totals."""
from __future__ import annotations

from dataclasses import replace
from datetime import date

from analyzer import SpendingReport, analyze_spending
from models import Transaction

# Matches default pace card cycle start; configured charges land on this day each month.
DEFAULT_BILLING_CYCLE_DAY = 10


def fixed_charge_date_for_billing(
    billing_date: date,
    cycle_day: int = DEFAULT_BILLING_CYCLE_DAY,
) -> date:
    """Pin recurring charges to the billing cycle start (e.g. 10th), not statement close date."""
    import calendar

    day = min(max(cycle_day, 1), 28)
    last = calendar.monthrange(billing_date.year, billing_date.month)[1]
    return date(billing_date.year, billing_date.month, min(day, last))


def _normalize_charge(charge: dict) -> dict:
    schedule = charge.get("schedule", "monthly")
    if schedule not in ("monthly", "once"):
        schedule = "monthly"
    from_m = str(charge.get("from_month", "")).strip()
    through_m = str(charge.get("through_month", "")).strip()
    charge_date = charge.get("charge_date")
    if charge_date:
        charge_date = str(charge_date).strip()[:10]
    if schedule == "once" and charge_date:
        ym = charge_date[:7]
        from_m = ym
        through_m = ym
    out = {
        "id": str(charge["id"]).strip(),
        "name_en": str(charge["name_en"]).strip(),
        "name_he": charge.get("name_he"),
        "amount": round(float(charge["amount"]), 2),
        "category_en": str(charge.get("category_en", "Uncategorized")).strip(),
        "from_month": from_m,
        "through_month": through_m,
        "schedule": schedule,
    }
    if charge_date:
        out["charge_date"] = charge_date
    return out


def _load_charges() -> list[dict]:
    """Fixed charges are applied by the Node API from Supabase, not local files."""
    return []


def month_key(billing_date: date | str | None) -> str | None:
    if billing_date is None:
        return None
    if isinstance(billing_date, str):
        return billing_date[:7]
    return billing_date.strftime("%Y-%m")


def cycle_start_for_statement_billing(
    billing_date: date,
    cycle_day: int = DEFAULT_BILLING_CYCLE_DAY,
) -> date:
    """Which cycle a statement bill belongs to (matches web pace utils)."""
    day = min(max(cycle_day, 1), 28)
    if billing_date.month == 1:
        prev_year = billing_date.year - 1
        prev_month = 12
    else:
        prev_year = billing_date.year
        prev_month = billing_date.month - 1
    import calendar

    last = calendar.monthrange(prev_year, prev_month)[1]
    return date(prev_year, prev_month, min(day, last))


def cycle_month_key_for_billing(billing_date: date | str | None) -> str | None:
    """Billing-cycle month (e.g. Feb 2026), not the statement charge month."""
    parsed = _parse_billing_date(billing_date)
    if parsed is None:
        return None
    return cycle_start_for_statement_billing(parsed).strftime("%Y-%m")


def charges_for_billing(billing_date: date | str | None) -> list[dict]:
    ym = cycle_month_key_for_billing(billing_date)
    if not ym:
        return []
    applicable: list[dict] = []
    for charge in _load_charges():
        from_m = charge.get("from_month", "")
        through_m = charge.get("through_month")
        if through_m and from_m <= ym <= through_m:
            applicable.append(charge)
    return applicable


def _parse_billing_date(billing: date | str | None) -> date | None:
    if billing is None:
        return None
    if isinstance(billing, date):
        return billing
    return date.fromisoformat(billing)


def augment_report(report: SpendingReport) -> SpendingReport:
    from exclusions import apply_exclusions

    billing_date = _parse_billing_date(report.metadata.get("billing_date"))
    if billing_date is None:
        return apply_exclusions(report)

    applicable = charges_for_billing(billing_date)
    cycle_start = cycle_start_for_statement_billing(billing_date)
    charge_by_id = {c["id"]: c for c in applicable}
    month_label = billing_date.strftime("%b %Y")

    updated_txs: list[Transaction] = []
    for tx in report.transactions:
        if tx.notes and tx.notes.startswith("fixed_charge:"):
            charge_id = tx.notes.split(":", 1)[1]
            charge = charge_by_id.get(charge_id)
            if not charge:
                continue
            if charge.get("schedule") == "once" and charge.get("charge_date"):
                charge_date = date.fromisoformat(charge["charge_date"])
            else:
                charge_date = fixed_charge_date_for_billing(cycle_start)
            updated_txs.append(
                replace(
                    tx,
                    date=charge_date,
                    merchant_he=str(charge.get("name_he", charge["name_en"])),
                    merchant_en=str(charge["name_en"]),
                    amount=float(charge["amount"]),
                    charge_amount=float(charge["amount"]),
                    category_en=str(charge.get("category_en", "Uncategorized")),
                    billing_month=month_label,
                )
            )
            continue
        updated_txs.append(tx)

    existing_ids = {
        tx.notes.split(":", 1)[1]
        for tx in updated_txs
        if tx.notes and tx.notes.startswith("fixed_charge:")
    }

    new_txs: list[Transaction] = []
    for charge in applicable:
        charge_id = charge["id"]
        if charge_id in existing_ids:
            continue
        if charge.get("schedule") == "once" and charge.get("charge_date"):
            tx_date = date.fromisoformat(charge["charge_date"])
            tx_type = "חיוב ידני"
        else:
            tx_date = fixed_charge_date_for_billing(cycle_start)
            tx_type = "חיוב קבוע"
        new_txs.append(
            Transaction(
                date=tx_date,
                merchant_he=str(charge.get("name_he", charge["name_en"])),
                amount=float(charge["amount"]),
                charge_amount=float(charge["amount"]),
                transaction_type_he=tx_type,
                category_he=None,
                notes=f"fixed_charge:{charge_id}",
                merchant_en=str(charge["name_en"]),
                category_en=str(charge.get("category_en", "Uncategorized")),
                merchant_known=True,
                billing_month=month_label,
            )
        )

    if not new_txs and updated_txs == list(report.transactions):
        return apply_exclusions(report)

    result = analyze_spending(updated_txs + new_txs, dict(report.metadata))
    return apply_exclusions(result)

File: api/src/services/fixedCharges.ts
import { readFileSync } from "fs";
import path from "path";
import type { FixedCharge, FixedChargesData, SpendingReport, Transaction } from "../types.js";
import { readFixedCharges, writeFixedCharges } from "../storage/index.js";
import { monthLabelFromIso } from "../utils/dates.js";

const BUILTIN_PATH = path.resolve(
  process.env.DATA_DIR ? path.dirname(process.env.DATA_DIR) : path.join(process.cwd(), ".."),
  "data",
  "fixed_charges.json",
);

const ONGOING_THROUGH_MONTH = "2035-12";
const MONTH_RE = /^\d{4}-\d{2}$/;
const DATE_RE = /^\d{4}-\d{2}-\d{2}$/;

function isOneTime(charge: FixedCharge): boolean {
  return charge.schedule === "once";
}

/** Matches default pace card cycle start; configured charges land on this day each month. */
export const DEFAULT_BILLING_CYCLE_DAY = 10;

/** Inverse of statement billing date — cycle start for that statement (matches web pace utils). */
export function cycleStartForStatementBilling(billingDate: string, cycleDay = DEFAULT_BILLING_CYCLE_DAY): string {
  const [y, m] = billingDate.slice(0, 10).split("-").map(Number);
  const prev = new Date(y, m - 2, cycleDay);
  const day = Math.min(Math.max(cycleDay, 1), 28);
  const lastDay = new Date(prev.getFullYear(), prev.getMonth() + 1, 0).getDate();
  const safeDay = Math.min(day, lastDay);
  return `${prev.getFullYear()}-${String(prev.getMonth() + 1).padStart(2, "0")}-${String(safeDay).padStart(2, "0")}`;
}

/** Billing-cycle month key (what users see as "Feb 2026"), not the statement charge month. */
export function cycleMonthKeyForBilling(billingDate: string, cycleDay = DEFAULT_BILLING_CYCLE_DAY): string {
  return cycleStartForStatementBilling(billingDate, cycleDay).slice(0, 7);
}
/** Pin recurring charges to the billing cycle start (e.g. 10th), not the statement close date. */
export function fixedChargeDateForBilling(
  billingDate: string,
  cycleDay = DEFAULT_BILLING_CYCLE_DAY,
): string {
  const [y, m] = billingDate.slice(0, 7).split("-").map(Number);
  const day = Math.min(Math.max(cycleDay, 1), 28);
  const lastDay = new Date(y, m, 0).getDate();
  const safeDay = Math.min(day, lastDay);
  return `${y}-${String(m).padStart(2, "0")}-${String(safeDay).padStart(2, "0")}`;
}

let cached: FixedCharge[] | null = null;

function loadBuiltinCharges(): FixedCharge[] {
  try {
    const raw = readFileSync(BUILTIN_PATH, "utf-8");
    const data = JSON.parse(raw) as { charges?: FixedCharge[] };
    return (data.charges || []).map(normalizeCharge);
  } catch {
    return [];
  }
}

function normalizeCharge(charge: FixedCharge): FixedCharge {
  const schedule = charge.schedule === "once" ? "once" : "monthly";
  const amount = Math.round(charge.amount * 100) / 100;
  let fromMonth = charge.from_month.trim();
  let throughMonth = charge.through_month.trim();
  let chargeDate = charge.charge_date?.trim().slice(0, 10);

  if (schedule === "once") {
    if (chargeDate && DATE_RE.test(chargeDate)) {
      const ym = chargeDate.slice(0, 7);
      fromMonth = ym;
      throughMonth = ym;
    }
  }

  return {
    id: charge.id.trim(),
    name_en: charge.name_en.trim(),
    name_he: charge.name_he?.trim() || undefined,
    amount,
    category_en: charge.category_en.trim(),
    from_month: fromMonth,
    through_month: throughMonth,
    schedule,
    charge_date: chargeDate,
  };
}

function mergeCharges(user: FixedChargesData): FixedCharge[] {
  if (user.charges?.length) return user.charges.map(normalizeCharge);
  return loadBuiltinCharges();
}

export async function refreshFixedChargesCache(): Promise<void> {
  const user = await readFixedCharges();
  cached = mergeCharges(user);
}

function ensureCache(): void {
  if (!cached) cached = loadBuiltinCharges();
}

export function loadFixedCharges(): FixedCharge[] {
  ensureCache();
  return cached!;
}

export function configuredChargesForMonth(billingYm: string): FixedCharge[] {
  const byId = new Map<string, FixedCharge>();
  for (const charge of loadFixedCharges()) {
    if (charge.from_month <= billingYm && billingYm <= charge.through_month) {
      byId.set(charge.id, charge);
    }
  }
  return [...byId.values()];
}

export function validateFixedCharges(charges: FixedCharge[]): string | null {
  for (const raw of charges) {
    const charge = normalizeCharge(raw);
    if (!charge.id) return "Each charge needs an id";
    if (!charge.name_en) return "Each charge needs an English name";
    if (!charge.category_en) return "Each charge needs a category";
    if (!Number.isFinite(charge.amount) || charge.amount <= 0) return "Amount must be positive";
    if (isOneTime(charge)) {
      if (!charge.charge_date || !DATE_RE.test(charge.charge_date)) {
        return `${charge.name_en}: one-time charges need a date (YYYY-MM-DD)`;
      }
      continue;
    }
    if (!MONTH_RE.test(charge.from_month) || !MONTH_RE.test(charge.through_month)) {
      return "Months must be YYYY-MM";
    }
    if (charge.from_month > charge.through_month) {
      return `${charge.name_en}: start month must be on or before end month`;
    }
  }
  return null;
}

export async function saveFixedCharges(charges: FixedCharge[]): Promise<FixedCharge[]> {
  const normalized = charges.map(normalizeCharge);
  const error = validateFixedCharges(normalized);
  if (error) throw new Error(error);
  await writeFixedCharges({ charges: normalized });
  cached = normalized;
  return normalized;
}


function chargesForBilling(billingDate: string | undefined, charges: FixedCharge[]): FixedCharge[] {
  const ym = billingDate ? cycleMonthKeyForBilling(billingDate) : null;
  if (!ym) return [];
  return charges.filter((c) => c.from_month <= ym && ym <= c.through_month);
}

function mergeCategorySummaries(report: SpendingReport): SpendingReport["by_category"] {
  const totals = new Map<string, { total: number; count: number; he: string | null }>();
  let grand = 0;
  for (const tx of report.transactions) {
    const cat = tx.category_en || "Uncategorized";
    grand += tx.charge_amount;
    const cur = totals.get(cat) || { total: 0, count: 0, he: tx.category_he };
    cur.total += tx.charge_amount;
    cur.count += 1;
    totals.set(cat, cur);
  }
  return [...totals.entries()]
    .map(([category_en, v]) => ({
      category_en,
      category_he: v.he,
      total: v.total,
      count: v.count,
      share_pct: grand ? (v.total / grand) * 100 : 0,
    }))
    .sort((a, b) => b.total - a.total);
}

function topMerchants(report: SpendingReport): SpendingReport["top_merchants"] {
  const totals = new Map<string, { total: number; count: number; he: string; cat: string }>();
  for (const tx of report.transactions) {
    const key = tx.merchant_en || tx.merchant_he;
    const cur = totals.get(key) || {
      total: 0,
      count: 0,
      he: tx.merchant_he,
      cat: tx.category_en,
    };
    cur.total += tx.charge_amount;
    cur.count += 1;
    totals.set(key, cur);
  }
  return [...totals.entries()]
    .map(([merchant_en, v]) => ({
      merchant_en,
      merchant_he: v.he,
      category_en: v.cat,
      total: v.total,
      count: v.count,
    }))
    .sort((a, b) => b.total - a.total)
    .slice(0, 15);
}

export function augmentReport(report: SpendingReport): SpendingReport {
  const billingDate = report.metadata.billing_date as string | undefined;
  if (!billingDate) return report;

  const charges = chargesForBilling(billingDate, loadFixedCharges());
  if (!charges.length) return report;

  const cycleStart = cycleStartForStatementBilling(billingDate);
  const chargeById = new Map(charges.map((c) => [c.id, c]));
  const label = monthLabelFromIso(billingDate);

  const transactions = report.transactions.map((tx) => {
    if (!tx.notes?.startsWith("fixed_charge:")) return tx;
    const id = tx.notes.slice("fixed_charge:".length);
    const charge = chargeById.get(id);
    if (!charge) return tx;
    const chargeDate = isOneTime(charge) && charge.charge_date
      ? charge.charge_date
      : fixedChargeDateForBilling(cycleStart);
    return {
      ...tx,
      date: chargeDate,
      merchant_he: charge.name_he || charge.name_en,
      merchant_en: charge.name_en,
      amount: charge.amount,
      charge_amount: charge.amount,
      category_en: charge.category_en,
      billing_month: label,
    };
  });

  const existingIds = new Set(
    transactions
      .filter((tx) => tx.notes?.startsWith("fixed_charge:"))
      .map((tx) => tx.notes!.slice("fixed_charge:".length)),
  );

  const newTxs: Transaction[] = [];
  for (const charge of charges) {
    if (existingIds.has(charge.id)) continue;
    const chargeDate = isOneTime(charge) && charge.charge_date
      ? charge.charge_date
      : fixedChargeDateForBilling(cycleStart);
    newTxs.push({
      date: chargeDate,
      merchant_he: charge.name_he || charge.name_en,
      merchant_en: charge.name_en,
      amount: charge.amount,
      charge_amount: charge.amount,
      transaction_type_he: isOneTime(charge) ? "חיוב ידני" : "חיוב קבוע",
      category_he: null,
      category_en: charge.category_en,
      notes: `fixed_charge:${charge.id}`,
      merchant_known: true,
      billing_month: label,
    });
  }

  if (!newTxs.length && transactions.every((tx, i) => tx === report.transactions[i])) return report;

  const merged = [...transactions, ...newTxs].sort((a, b) => b.date.localeCompare(a.date));
  const dates = merged.map((t) => t.date).sort();
  const total = merged.reduce((s, t) => s + t.charge_amount, 0);
  const augmented: SpendingReport = {
    ...report,
    transactions: merged,
    total_spent: total,
    transaction_count: transactions.length,
    date_range: dates.length ? [dates[0], dates[dates.length - 1]] : report.date_range,
    by_category: [],
    top_merchants: [],
    unknown_merchants: report.unknown_merchants.filter(
      (m) => !newTxs.some((tx) => tx.merchant_he === m),
    ),
  };
  augmented.by_category = mergeCategorySummaries(augmented);
  augmented.top_merchants = topMerchants(augmented);
  return augmented;
}

/** Recompute totals and category/merchant summaries from current transactions. */
export function rebuildReportSummaries(report: SpendingReport): SpendingReport {
  const txs = report.transactions;
  const dates = txs.map((t) => t.date).sort();
  const total = txs.reduce((s, t) => s + t.charge_amount, 0);
  const updated: SpendingReport = {
    ...report,
    total_spent: total,
    transaction_count: txs.length,
    date_range: dates.length ? [dates[0], dates[dates.length - 1]] : report.date_range,
    by_category: [],
    top_merchants: [],
  };
  updated.by_category = mergeCategorySummaries(updated);
  updated.top_merchants = topMerchants(updated);
  return updated;
}

export { ONGOING_THROUGH_MONTH };
